File: BEmu/BloombergTypes/Subscription.cs
//------------------------------------------------------------------------------
//  This code comes from Bloomberg.Blpapi.dll
//------------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using System.Text.RegularExpressions;

namespace Bloomberglp.Blpapi
{
    public class Subscription
    {
        #region NOT PART OF DLL
        private readonly string _security;
        internal string Security { get { return this._security; } }

        private readonly List<string> _fields;
        internal List<string> Fields { get { return this._fields; } }

        private static string CheckSecurity(string security)
        {
            if (security.Contains(TOPIC_OPTIONS_SEPARATOR))
                throw new ArgumentNullException(security);
            return security;
        }

        private static string CheckField(string field)
        {
            if (field.Contains(OPTIONS_SEPARATOR))
                throw new ArgumentException(field);
            return field;
        }

        private int? ReadConflationInterval(IList<string> options)
        {
            int? result = null;
            foreach (var item in options)
            {
                if (item.StartsWith("interval="))
                {
                    int temp;
                    string strInterval = item.Substring(item.IndexOf('=') + 1);
                    if (int.TryParse(strInterval, out temp))
                        result = temp;
                }
            }
            return result;
        }

        private readonly int? _conflationInterval;
        internal int? ConflationInterval { get { return this._conflationInterval; } }
        #endregion

        private const char OPTIONS_SEPARATOR = '&';
		private const char TOPIC_OPTIONS_SEPARATOR = '?';
		private const char FIELDS_SEPARATOR = ',';
		private const string FIELDS_OPTION = "fields=";
		private string d_subscriptionString;
		private CorrelationID d_correlationId;
		private Session.SubscriptionStatus d_subscriptionStatus = Session.SubscriptionStatus.UNSUBSCRIBED;
		private bool d_isResolved;
		private static readonly Regex utsReg = new Regex("^//[^/]+/[^/]+/[^?]+([?](.)+)?$", RegexOptions.None);
		private static readonly Regex topicReg = new Regex("^[^?]+([?](.)+)?$", RegexOptions.None);
		public string SubscriptionString
		{
			get
			{
				return this.d_subscriptionString;
			}
			set
			{
				if (value == null)
				{
					throw new ArgumentNullException();
				}
				if (value.Length <= 0)
				{
					throw new ArgumentException("An empty subscription string");
				}
				this.d_subscriptionString = value;
			}
		}
		public CorrelationID CorrelationID
		{
			get
			{
				return this.d_correlationId;
			}
			set
			{
				if (value == null)
				{
					throw new ArgumentNullException();
				}
				this.d_correlationId = value;
			}
		}
		public Session.SubscriptionStatus SubscriptionStatus
		{
			get
			{
				return this.d_subscriptionStatus;
			}
		}
		public bool IsResolved
		{
			get
			{
				return this.d_isResolved;
			}
		}
		internal Subscription(string subscriptionString, CorrelationID correlationID, Session.SubscriptionStatus status)
		{
			this.d_subscriptionString = subscriptionString;
			this.d_correlationId = correlationID;
			this.d_subscriptionStatus = status;
		}
		public Subscription(string subscriptionString) : this(subscriptionString, new CorrelationID())
		{
		}
		public Subscription(string subscriptionString, CorrelationID correlationID) : this(subscriptionString, correlationID, false)
		{
		}
		public Subscription(string subscriptionString, bool isResolved) : this(subscriptionString, new CorrelationID(), isResolved)
		{
		}
		public Subscription(string subscriptionString, CorrelationID correlationID, bool isResolved)
		{
			if (subscriptionString == null || correlationID == null)
			{
				throw new ArgumentNullException();
			}
			if (!Subscription.isValidSubscriptionString(subscriptionString))
			{
				throw new ArgumentException("An empty subscription string");
			}
			this.d_subscriptionString = subscriptionString;
			this.d_correlationId = correlationID;
			this.d_isResolved = isResolved;

            //set security and fields
		}
		public Subscription(string security, string fields) : this(security, fields, "", new CorrelationID())
		{
		}
		public Subscription(string security, string fields, CorrelationID correlationId) : this(security, fields, "", correlationId)
        {
		}
		public Subscription(string security, string fields, string options) : this(security, fields, options, new CorrelationID())
        {
		}
		public Subscription(string security, string fields, string options, CorrelationID correlationID)
		{
            if (security == null || security.Length <= 0 || security.IndexOf(TOPIC_OPTIONS_SEPARATOR) >= 0)
			{
				throw new ArgumentNullException("Invalid security: " + security);
			}
			if (options == null)
			{
				options = "";
			}
			if (fields == null)
			{
				fields = "";
			}
            if (fields.IndexOf(OPTIONS_SEPARATOR) >= 0)
			{
				throw new ArgumentException("Invalid fields: " + fields);
			}
			if (correlationID == null)
			{
				correlationID = new CorrelationID();
			}
			StringBuilder stringBuilder = new StringBuilder(security.Length + fields.Length + options.Length);
			stringBuilder.Append(security);
			bool flag = fields.Length > 0;
			bool flag2 = options.Length > 0;
			if (flag || flag2)
			{
                stringBuilder.Append(TOPIC_OPTIONS_SEPARATOR);
				if (flag)
				{
                    stringBuilder.Append(FIELDS_OPTION);
					stringBuilder.Append(fields);
				}
				if (flag && flag2)
				{
                    stringBuilder.Append(OPTIONS_SEPARATOR);
				}
				if (flag2)
				{
					stringBuilder.Append(options);
				}
			}
			this.d_subscriptionString = stringBuilder.ToString();
			this.d_correlationId = correlationID;



            this._security = security;
            this._fields = new List<string>(fields.Split(FIELDS_SEPARATOR).Select(s => Subscription.CheckField(s)));
            this._conflationInterval = this.ReadConflationInterval(new List<string>(options.Split(FIELDS_SEPARATOR)));
		}
		public Subscription(string security, IList<string> fields) : this(security, fields, null, new CorrelationID())
		{
		}
		public Subscription(string security, IList<string> fields, CorrelationID correlationID) : this(security, fields, null, correlationID)
		{
		}
		public Subscription(string security, IList<string> fields, IList<string> options) : this(security, fields, options, new CorrelationID())
		{
		}
        public Subscription(string security, IList<string> fields, IList<string> options, CorrelationID correlationID)
		{
            if (security == null || security.Length <= 0 || security.IndexOf(TOPIC_OPTIONS_SEPARATOR) >= 0)
			{
				throw new ArgumentNullException("Invalid security: " + security);
			}
			if (correlationID == null)
			{
				throw new ArgumentException("Null argument");
			}
			StringBuilder stringBuilder = new StringBuilder(64);
			stringBuilder.Append(security);
			bool hasFields = fields != null && fields.Count != 0;
			bool hasOptions = options != null && options.Count != 0;
			bool separatorsAppended = false;
			if (hasFields)
			{
				foreach (string current in fields)
				{
					if (current != null && current.Length > 0)
					{
						if (!separatorsAppended)
						{
                            stringBuilder.Append(TOPIC_OPTIONS_SEPARATOR);
                            stringBuilder.Append(FIELDS_OPTION);
							separatorsAppended = true;
						}
						else
						{
                            stringBuilder.Append(FIELDS_SEPARATOR);
						}
						stringBuilder.Append(current);
					}
				}
			}
			if (hasOptions)
			{
				foreach (string current2 in options)
				{
					if (current2 != null && current2.Length > 0)
					{
						if (!separatorsAppended)
						{
                            stringBuilder.Append(TOPIC_OPTIONS_SEPARATOR);
							separatorsAppended = true;
						}
						else
						{
                            stringBuilder.Append(OPTIONS_SEPARATOR);
						}
						stringBuilder.Append(current2);
					}
				}
			}
			this.d_subscriptionString = stringBuilder.ToString();
			this.d_correlationId = correlationID;
            this._security = Subscription.CheckSecurity(security);

            if (fields != null)
                this._fields = new List<string>(fields.Select(s => Subscription.CheckField(s)));

            if (options != null)
                this._conflationInterval = this.ReadConflationInterval(options);
		}
		public override string ToString()
		{
			StringBuilder stringBuilder = new StringBuilder();
			stringBuilder.Append("[");
			stringBuilder.Append(this.d_subscriptionString);
			stringBuilder.Append(" CorrelationID = {");
			stringBuilder.Append(this.d_correlationId.ToString());
			stringBuilder.Append("}");
			stringBuilder.Append(" Status = {");
			stringBuilder.Append(this.d_subscriptionStatus.ToString());
			stringBuilder.Append("}");
			stringBuilder.Append("]");
			return stringBuilder.ToString();
		}
		internal void setSubscriptionStatus(Session.SubscriptionStatus status)
		{
			this.d_subscriptionStatus = status;
		}
		internal static bool isValidSubscriptionString(string subscriptionString)
		{
			if (subscriptionString.StartsWith("//"))
			{
                return Subscription.utsReg.IsMatch(subscriptionString);
			}
            return Subscription.topicReg.IsMatch(subscriptionString);
		}
    }
}

File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/Subscription.java
//------------------------------------------------------------------------------
//	This code comes from blpapi-3.6.1-0.jar
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

public class Subscription
{
	// The following is not part of the DLL
	
    private final String _security;
    public String security()
    {
    	return this._security;
    }
    
    private final List<String> _fields;
    public List<String> fields()
    {
    	return this._fields;
    }
	
    private final Integer _conflationInterval;
    Integer conflationInterval() 
    {
    	return this._conflationInterval;
    }
	
    private Integer ReadConflationInterval(List<String> options)
    {
    	if(options == null)
    		return null;
    	
        Integer result = null;
        for(int i = 0; i < options.size(); i++)
        {
        	String str = options.get(i);
        	
            if (str.startsWith("interval="))
            {
                String strInterval = str.substring(0, str.indexOf('=') + 1);                
                if(this.tryParseInt(strInterval))
                {
                	result = Integer.parseInt(strInterval);
                }
            }
        }
        return result;
    }
    
    private boolean tryParseInt(String value)  
    {  
         try  
         {  
             Integer.parseInt(value);  
             return true;  
         }
         catch(NumberFormatException nfe)  
         {
        	 return false;  
         }  
    }
	// The above is not part of the DLL
	
	
	private String d_subscriptionString;
	private CorrelationID d_correlationId;
	private Session.SubscriptionStatus d_subscriptionStatus = Session.SubscriptionStatus.UNSUBSCRIBED;
	
	private static final String utsReg = "//[^/]+/[^/]+/[^?]+(\\?[^&]+(&[^&]+)*)?";
	private static final String topicReg = "[^?]+(\\?[^&]+(&[^&]+)*)?";

    private static final char OPTIONS_SEPARATOR = '&';
	private static final char TOPIC_OPTIONS_SEPARATOR = '?';
	private static final char FIELDS_SEPARATOR = ',';
	private static final String FIELDS_OPTION = "fields=";
	
	Subscription(String subscriptionString, CorrelationID correlationID, Session.SubscriptionStatus status)
	{
		this.d_subscriptionString = subscriptionString;
		this.d_correlationId = correlationID;
		this.d_subscriptionStatus = status;
		
		this._conflationInterval = null;
		this._security = null;
		this._fields = null;
	}
	
	public Subscription(String subscriptionString)
	{
		this(subscriptionString, new CorrelationID());
	}
	
	public Subscription(String subscriptionString, CorrelationID correlationID)
	{
		if ((subscriptionString == null) || (correlationID == null)) {
			throw new IllegalArgumentException("Null arguments");
		}
		if (!isValidSubscriptionString(subscriptionString)) {
			throw new IllegalArgumentException("Invalid subscription string");
		}
		this.d_subscriptionString = subscriptionString;
		this.d_correlationId = correlationID;
		
		this._conflationInterval = null;
		this._security = null;
		this._fields = null;
	}
	
	public Subscription(String security, String fields)
	{
		this(security, fields, "", new CorrelationID());
	}
	
	public Subscription(String security, String fields, CorrelationID correlationId)
	{
		this(security, fields, "", correlationId);
	}
	
	public Subscription(String security, String fields, String options)
	{
		this(security, fields, options, new CorrelationID());
	}
	
	public Subscription(String security, String fields, String options, CorrelationID correlationID)
	{
		if ((security == null) || (security.length() <= 0) || (security.indexOf(TOPIC_OPTIONS_SEPARATOR) >= 0)) {
			throw new IllegalArgumentException("Invalid security: " + security);
		}
		if (options == null) {
			options = "";
		}
		if (fields == null) {
			fields = "";
		}
		if (fields.indexOf(OPTIONS_SEPARATOR) >= 0) {
			throw new IllegalArgumentException("Invalid fields: " + fields);
		}
		if (correlationID == null) {
			correlationID = new CorrelationID();
		}
		StringBuilder localStringBuilder = new StringBuilder(security.length() + fields.length() + options.length());
		
		localStringBuilder.append(security);
		
		boolean hasFields = fields.length() > 0;
		boolean hasOptions = options.length() > 0; 
		
		if (hasFields || hasOptions)
		{
			localStringBuilder.append(TOPIC_OPTIONS_SEPARATOR);
			if (hasFields)
			{
				localStringBuilder.append(FIELDS_OPTION);
				localStringBuilder.append(fields);
			}
			
			if (hasFields && hasOptions)
			{
				localStringBuilder.append(OPTIONS_SEPARATOR);
			}
			if (hasOptions)
			{
				localStringBuilder.append(options);
			}
		}
		this.d_subscriptionString = localStringBuilder.toString();
		this.d_correlationId = correlationID;
		

		this._security = security;
		
		if(hasOptions)
		{
			String[] arrOptions = options.split(",");	
			this._conflationInterval = this.ReadConflationInterval(Arrays.asList(arrOptions));
		}
		else
			this._conflationInterval = null;
		
		if(hasFields)
		{
			String[] arrFields = fields.split(",");
			this._fields = Arrays.asList(arrFields);
		}
		else
			this._fields = null;
	}
	
	public Subscription(String security, List<String> fields)
	{
		this(security, fields, null, new CorrelationID());
	}
	
	public Subscription(String security, List<String> fields, CorrelationID correlationID)
	{
		this(security, fields, null, correlationID);
	}
	
	public Subscription(String security, List<String> fields, List<String> options)
	{
		this(security, fields, options, new CorrelationID());
	}
	
	public Subscription(String security, List<String> fields, List<String> options, CorrelationID correlationID)
	{
		if ((security == null) || (security.length() <= 0) || (security.indexOf(TOPIC_OPTIONS_SEPARATOR) >= 0)) {
			throw new IllegalArgumentException("Invalid security: " + security);
		}
		if (correlationID == null) {
			throw new IllegalArgumentException("Null correlationId argument");
		}
		StringBuilder localStringBuilder = new StringBuilder(64);
		
		localStringBuilder.append(security);
		
		getFieldsAndOptions(localStringBuilder, fields, options);
		
		this.d_subscriptionString = localStringBuilder.toString();
		this.d_correlationId = correlationID;
		
		this._security = security;
		this._conflationInterval = this.ReadConflationInterval(options);
		this._fields = fields;
	}
	
	public String subscriptionString()
	{
		return this.d_subscriptionString;
	}
	
	public CorrelationID correlationID()
	{
		return this.d_correlationId;
	}
	
	public Session.SubscriptionStatus subscriptionStatus()
	{
		return this.d_subscriptionStatus;
	}
	
	public void setSubscriptionString(String subscriptionString)
	{
		if (subscriptionString == null) {
			throw new IllegalArgumentException("A null argument");
		}
		if (subscriptionString.length() <= 0) {
			throw new IllegalArgumentException("An empty subscription string");
		}
		this.d_subscriptionString = subscriptionString;
	}
	
	public void setCorrelationID(CorrelationID correlationID)
	{
		if (correlationID == null) {
			throw new IllegalArgumentException("A null argument");
		}
		this.d_correlationId = correlationID;
	}
	
	public String toString()
	{
		StringBuilder localStringBuilder = new StringBuilder();
		
		localStringBuilder.append("[");
		
		localStringBuilder.append(this.d_subscriptionString);
		
		localStringBuilder.append(" CorrelationID = {");
		localStringBuilder.append(this.d_correlationId.toString());
		localStringBuilder.append("}");
		
		localStringBuilder.append(" Status = {");
		localStringBuilder.append(this.d_subscriptionStatus.toString());
		localStringBuilder.append("}");
		
		localStringBuilder.append("]");
		
		return localStringBuilder.toString();
	}
	
	private static void getFieldsAndOptions(StringBuilder stringBuilder, List<String> fields, List<String> options)
	{
		boolean hasFields = (fields != null) && (fields.size() != 0);
		boolean hasOptions = (options != null) && (options.size() != 0);
		boolean separatorsAppended = false;
		
		String current;
		
		Iterator<String> localIterator;
		if (hasFields) {
			for (localIterator = fields.iterator(); localIterator.hasNext();)
			{
				current = (String)localIterator.next();
				if ((current != null) && (current.length() > 0))
				{
					if (!separatorsAppended)
					{
						stringBuilder.append(TOPIC_OPTIONS_SEPARATOR);
						stringBuilder.append(FIELDS_OPTION);
						separatorsAppended = true;
					}
					else
					{
						stringBuilder.append(FIELDS_SEPARATOR);
					}
					stringBuilder.append(current);
				}
			}
		}
		
		if (hasOptions) {
			for (localIterator = options.iterator(); localIterator.hasNext();)
			{
				current = (String)localIterator.next();
				if ((current != null) && (current.length() > 0))
				{
					if (!separatorsAppended)
					{
						stringBuilder.append(TOPIC_OPTIONS_SEPARATOR);
						separatorsAppended = true;
					}
					else
					{
						stringBuilder.append(OPTIONS_SEPARATOR);
					}
					stringBuilder.append(current);
				}
			}
		}
	}
	
	static boolean isValidSubscriptionString(String paramString)
	{
		if (paramString.startsWith("//")) {
			return paramString.matches(utsReg);
		}
		return paramString.matches(topicReg);
	}
}
